libs/langchain-community/src/vectorstores/upstash.ts

```diff
@@ -0,0 +1,241 @@
+        id,
+        vector,
+        metadata,
+      };
+    });
+
+    const vectorChunks = chunkArray(upstashVectors, CONCURRENT_UPSERT_LIMIT);
+
+    const batchRequests = vectorChunks.map((chunk) =>
+      this.caller.call(async () => this.index.upsert(chunk))
+    );
+
+    await Promise.all(batchRequests);
+
+    return documentIds;
+  }
+
+  /**
+   * This method deletes documents from the Upstash database. You can either
+   * provide the target ids, or delete all vectors in the database.
+   * @param params Object containing either array of ids of the documents or boolean deleteAll.
+   * @returns Promise that resolves when the specified documents have been deleted from the database.
+   */
+  async delete(params: UpstashDeleteParams): Promise<void> {
+    if (params.deleteAll) {
+      await this.index.reset();
+    } else if (params.ids) {
+      await this.index.delete(params.ids);
+    } else {
+      throw new Error(
+        "You must provide either an array of target ids or deleteAll as true"
+      );
+    }
+  }
+
+  protected async _runUpstashQuery(
+    query: number[],
+    k: number,
+    options?: { includeVectors: boolean }
+  ) {
+    const queryResult = await this.index.query<UpstashQueryMetadata>({
+      vector: query,
+      topK: k,
+      includeMetadata: true,
+      ...options,
+    });
+
+    return queryResult;
+  }
+
+  /**
+   * This method performs a similarity search in the Upstash database
+   * over the existing vectors.
+   * @param query Query vector for the similarity search.
+   * @param k The number of similar vectors to return as result.
+   * @returns Promise that resolves with an array of tuples, each containing
+   *  Document object and similarity score. The length of the result will be
+   *  maximum of 'k' and vectors in the index.
+   */
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number
+  ): Promise<[Document, number][]> {
+    const results = await this._runUpstashQuery(query, k);
+
+    const searchResult: [Document, number][] = results.map((res) => {
+      const { pageContentLC, ...metadata } = (res.metadata ??
+        {}) as UpstashQueryMetadata;
+      return [
+        new Document({
+          metadata,
+          pageContent: pageContentLC,
+        }),
+        res.score,
+      ];
+    });
+
+    return searchResult;
+  }
+
+  /**
+   * This method creates a new UpstashVector instance from an array of texts.
+   * The texts are initially converted to Document instances and added to Upstash
+   * database.
+   * @param texts The texts to create the documents from.
+   * @param metadatas The metadata values associated with the texts.
+   * @param embeddings Embedding interface of choice, to create the text embeddings.
+   * @param dbConfig Object containing the Upstash database configs.
+   * @returns Promise that resolves with a new UpstashVector instance.
+   */
+  static async fromTexts(
+    texts: string[],
+    metadatas: object[] | object,
```

Comment:
Could we do this instead
```suggestion
    metadatas: UpstashMetadata[] | UpstashMetadata,
```

Reply:
How is `object` going to prevent something like this: `() => {}`. If the `Record` type in @upstash/vector is still preventing people pass funcs, arrays instead of `Record` I would say why not.